ggml-metal.metal

```diff
@@ -365,71 +365,80 @@ kernel void kernel_rms_norm(
     }
 }
 
+// putting them in the kernel cause a significant performance penalty
+#define N_DST 4 // each SIMD group works on 4 rows
+#define N_SIMDGROUP 2 // number of SIMD groups in a thread group
+#define N_SIMDWIDTH 32 // assuming SIMD group size is 32
 kernel void kernel_mul_mat_q4_0_f32(
         device const  void * src0,
         device const float * src1,
         device       float * dst,
         constant   int64_t & ne00,
         constant   int64_t & ne10,
         constant   int64_t & ne0,
-        threadgroup float  * sum [[threadgroup(0)]],
+        constant   int64_t & ne01[[buffer(4)]],
```

Comment:
What is the function of `[[buffer(4)]]` here?

Reply:
This gets the value from the buffer at index 4, corresponding the following line in `ggml-metal.m`:
`[encoder setBytes:&ne01 length:sizeof(ne01) atIndex:4];`